### PR TITLE
[Article] Network

### PR DESCRIPTION
**Pull request by NoNews**

### CODE REVIEW

**Comment by NoNews** [comment]
I'm driving some images at the moment. Feel free to review

File: docs/content/practices/network.md [hunk 1]
```diff
@@ -1,5 +1,148 @@
...
+* **Custom Interceptor**
+  <br>_In the simplest implementation, this is a common interceptor which is intercepts each http call, which is added only in tests and actually replaces
+  triggered requests for predefined responses. 
+  <br> However, unlike `OkHttpWebServer`, it works for more
+  high level, therefore, may miss some bugs. 
+  <br>Also, in the most of the cases, we don't need to test third party libraries code_
+* **DI**
+  <br> _Each DI framework ([Dagger2](https://github.com/google/dagger) / [Toothpick](https://github.com/stephanenicolas/toothpick) / [Koin](https://github.com/InsertKoinIO/koin) / etc) allows you to replace dependencies in tests.
+  <br> We can replace the `Retroift Service` or the whole` Repository` entirely. However, we actually crop
+  testing the entire network layer and offline mod if it exists_
```
Comment (sergio-sastre) [inline docs/content/practices/network.md:58]: typo: mode

File: docs/content/practices/network.md [hunk 2]
```diff
@@ -1,5 +1,148 @@
...
+
+**Problems**
+
+You need to mock manually. Imagine, that in your flow has hundreds of http requests.
+<br> You may spend a lot of time to have it done, and it's may be really challenging to keep these tests up-to-dated.
```
Comment (sergio-sastre) [inline docs/content/practices/network.md:63]: typos:
maybe
up-to-date

Comment (sergio-sastre) [reply docs/content/practices/network.md:63]: a little bit more of emphasis on the issue:
```suggestion
<br> You may spend a lot of time mocking every single http response, and it's really challenging to keep these tests up-to-date.
```

File: docs/content/practices/network.md [hunk 3]
```diff
@@ -1,5 +1,148 @@
...
+
+The key idea is that two modes of network operation appear in the tests:`Record & Replay`.
+
+* **Record**
+  <br>_In this mode, the test uses a real backend. However, it also records all requests and exports them to set or to one file that can be associated with a specific test. 
+  <br>This mode is used locally only when writing a test_
+
+* **Replay**
+  <br>_Instead of a real network, an already recorded file is used. This mode is used both locally and on CI. It allows to fully exclude all the possible problems with a network.
+  Also, we've got an opportunity to easily mock authorisation in this mode, as we've already had all the data we need to put into preferences/stores_
```
Comment (sergio-sastre) [inline docs/content/practices/network.md:75]: typo: as we already have

Comment (sergio-sastre) [reply docs/content/practices/network.md:75]: Also not very obvious what this "all the data we need to put into preferences/stores" means.

I suggest something like:
We already have retrieved user's profile data. This data is likely stored locally in disk. In doing so, we can save in http requests that require such infos by fetching user's profile data swiftly from the disk instead

File: docs/content/practices/network.md [hunk 4]
```diff
@@ -1,5 +1,148 @@
...
+* Account preparation
+  <br>_You need to prepare a user which matches the special condition you need to test. It can be still mocked manually or via specially created service, which does this job for you_
+  <br>
+* Core changes
+  <br>_As soon as your root code base changed in terms of http requests, you probably would need to re-record some tests. That's why it's better to skip authorisation mode in the `playback` mode, to reduce such cases_
+  
+
+### Conclusion
+
+There is no perfect solution exists, it all depends on a number of factors, for example, the size of the team and how often the code base changes.
```
Comment (sergio-sastre) [inline docs/content/practices/network.md:124]: <del>There is </del> no perfect solution exists

File: docs/content/practices/network.md [hunk 5]
```diff
@@ -1,5 +1,148 @@
...
+It also depends on what goals you set in UI testing.
+
+??? question "Should I avoid real network testing in general?"
+    Depends on UI testing goals, project type and testing priorities.
+    <br> <br> If it's important for you to have UI testing as a part of `Pull Request` or `Release` pipelines, then it's very important to put the stability of these tests first
```
Comment (sergio-sastre) [inline docs/content/practices/network.md:129]: I would add some enumeration here since these are very good points and need some highlighting

**Review by sergio-sastre — COMMENTED** [review]
I just have small nits and some wording change suggestions to make the ideas more clear. All in all, I liked it :)

File: docs/content/practices/network.md [hunk 6]
```diff
@@ -1,5 +1,148 @@
...
+### Real network
+
+* **Production server**
+  <br>_You will get 100% `E2E`. On the other hand, not all applications will be able to do this (for example, in the banking community)_
+
+* **Development server**
+  <br>_We still have `E2E`, but the environment can be significantly different from production. Also, regular redeployment
+  can be a problem_
+* **Special stage server**
+  <br>_Can be different with your production/development backend. Also pay your attention, that it requires an
```
[the remark this answers is not in the document]
Comment (NoNews) [reply docs/content/practices/network.md:36]: fixed

File: docs/content/practices/network.md [hunk 7]
```diff
@@ -1,5 +1,148 @@
...
+  additional effort from engineers to support it. It also is not 100% end to end, as it tests not real backend side._
+  
+???+ warning
+    It doesn't matter which of the options above you choose: Internet connection always can be an issue.
```
[the remark this answers is not in the document]
Comment (NoNews) [reply docs/content/practices/network.md:40]: fixed

File: docs/content/practices/network.md [hunk 8]
```diff
@@ -1,5 +1,148 @@
...
+* **Development server**
+  <br>_We still have `E2E`, but the environment can be significantly different from production. Also, regular redeployment
+  can be a problem_
+* **Special stage server**
+  <br>_Can be different with your production/development backend. Also pay your attention, that it requires an
+  additional effort from engineers to support it. It also is not 100% end to end, as it tests not real backend side._
+  
+???+ warning
+    It doesn't matter which of the options above you choose: Internet connection always can be an issue. 
+    <br>If you use UI testing as a part of pull request or release pipeline, it's better to avoid real network usage there, but just run them regular several times a day
```
[the remark this answers is not in the document]
Comment (NoNews) [reply docs/content/practices/network.md:41]: fixed

File: docs/content/practices/network.md (hunk 1 above)
Comment (NoNews) [reply docs/content/practices/network.md:58]: fixed

File: docs/content/practices/network.md (hunk 4 above)
Comment (NoNews) [reply docs/content/practices/network.md:124]: fixed

File: docs/content/practices/network.md (hunk 3 above)
Comment (NoNews) [reply docs/content/practices/network.md:75]: fixed

File: docs/content/practices/network.md [hunk 9]
```diff
@@ -1,5 +1,148 @@
...
+### Problems
+
+Ideally, we want to achieve the next picture:
+
+<Picture1 TBD>
+
+As we can see, all levels of the application are covered, the test is fully integrated, therefore, more reliable, with
+a product quality point of view. If the backend changed the contract, we will know about it right away.
+
+Unfortunately, this has a downside: everything about the network is unstable. The internet suddenly works a lot of worse or stop working altogether, or your backend may have redeployment step.
```
[the remark this answers is not in the document]
Comment (NoNews) [reply docs/content/practices/network.md:16]: good one. Added

File: docs/content/practices/network.md [hunk 10]
```diff
@@ -1,5 +1,148 @@
...
+
+
+All of this can cause your test to fail and show a non-representative false result. On the other side,
+if you completely abandon the network, you cannot name it completely
```
[the remark this answers is not in the document]
Comment (NoNews) [reply docs/content/practices/network.md:20]: fixed

File: docs/content/practices/network.md [hunk 11]
```diff
@@ -1,5 +1,148 @@
 # Network
 
-To be done
+# Network
+
+UI testing can play a role of regular auto-testing in your project, which checks communication of your feature with other features on the device. In that case, you have to choose, how to deal with a network.
```
[the remark this answers is not in the document]
Comment (NoNews) [reply docs/content/practices/network.md:5]: fixed, good one

[the remark this answers is not in the document]
Comment (NoNews) [reply docs/content/practices/network.md:5]: fixed, good one

**Review by v1sar — APPROVED** [review]
good

**Pull request merged**